controllers/alert_controller.go

```diff
@@ -99,95 +99,66 @@ func (r *AlertReconciler) SetupWithManagerAndOptions(mgr ctrl.Manager, opts Aler
+		// Record Prometheus metrics.
+		r.Metrics.RecordReadiness(ctx, obj)
+		r.Metrics.RecordDuration(ctx, obj, reconcileStart)
+		r.Metrics.RecordSuspend(ctx, obj, obj.Spec.Suspend)
 
-			readyCondition := conditions.Get(alert, meta.ReadyCondition)
-			switch readyCondition.Status {
-			case metav1.ConditionFalse:
-				// As we are no longer reconciling and the end-state is not ready, the reconciliation has stalled
-				conditions.MarkStalled(alert, readyCondition.Reason, readyCondition.Message)
-			case metav1.ConditionTrue:
-				// As we are no longer reconciling and the end-state is ready, the reconciliation is no longer stalled
-				conditions.Delete(alert, meta.StalledCondition)
-			}
-		}
-
-		if err := patchHelper.Patch(ctx, alert, patchOpts...); err != nil {
-			retErr = kerrors.NewAggregate([]error{retErr, err})
-		}
-
-		r.Metrics.RecordReadiness(ctx, alert)
-		r.Metrics.RecordDuration(ctx, alert, start)
+		// Patch finalizers, status and conditions.
+		retErr = r.patch(ctx, obj, patcher)
 	}()
 
-	if !controllerutil.ContainsFinalizer(alert, v1beta1.NotificationFinalizer) {
-		controllerutil.AddFinalizer(alert, v1beta1.NotificationFinalizer)
+	if !controllerutil.ContainsFinalizer(obj, apiv1.NotificationFinalizer) {
+		controllerutil.AddFinalizer(obj, apiv1.NotificationFinalizer)
 		result = ctrl.Result{Requeue: true}
 		return
 	}
 
-	if !alert.ObjectMeta.DeletionTimestamp.IsZero() {
-		controllerutil.RemoveFinalizer(alert, v1beta1.NotificationFinalizer)
+	if !obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		controllerutil.RemoveFinalizer(obj, apiv1.NotificationFinalizer)
 		result = ctrl.Result{}
 		return
 	}
 
-	return r.reconcile(ctx, alert)
+	// Return early if the object is suspended.
+	if obj.Spec.Suspend {
+		log.Info("Reconciliation is suspended for this object")
+		return ctrl.Result{}, nil
+	}
+
+	return r.reconcile(ctx, obj)
 }
 
-func (r *AlertReconciler) reconcile(ctx context.Context, alert *v1beta1.Alert) (ctrl.Result, error) {
+func (r *AlertReconciler) reconcile(ctx context.Context, alert *apiv1.Alert) (ctrl.Result, error) {
 	// Mark the resource as under reconciliation
-	conditions.MarkReconciling(alert, meta.ProgressingReason, "")
+	conditions.MarkReconciling(alert, meta.ProgressingReason, "Reconciliation in progress")
 
 	// validate alert spec and provider
 	if err := r.validate(ctx, alert); err != nil {
-		conditions.MarkFalse(alert, meta.ReadyCondition, v1beta1.ValidationFailedReason, err.Error())
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+		conditions.MarkFalse(alert, meta.ReadyCondition, apiv1.ValidationFailedReason, err.Error())
+		return ctrl.Result{Requeue: true}, client.IgnoreNotFound(err)
 	}
 
-	conditions.MarkTrue(alert, meta.ReadyCondition, meta.SucceededReason, v1beta1.InitializedReason)
+	conditions.MarkTrue(alert, meta.ReadyCondition, meta.SucceededReason, apiv1.InitializedReason)
 	ctrl.LoggerFrom(ctx).Info("Alert initialized")
```

Comment:
None of the reconcilers have a K8s event recorder. Is there any particular reason for that?
I'd expect such logs about object being ready to be also emitted as k8s event on the object.
In source-controller, we have a helper function to log and emit event at the same time https://github.com/fluxcd/source-controller/blob/ade77ec95abef43286fc3e2fc3c37e4e613a1ce3/controllers/gitrepository_controller.go#L939-L948 because we do the same multiple times.
Since these are really small reconcilers, there's no need of such helpers but it'd be good to have all the flux object behave similarly and show in the k8s events what's happening to the object, success or failure messages.

Reply:
Based on the same idea, when the validation fails above, a warning event can be emitted. Likewise in all the other reconcilers, wherever they fail during any operation.

Again, all these will increase k8s events. We can discuss if we should provide an option to skip certain events from controllers.

Reply:
Thanks. But it's only the failure event. What about the success event like we do for other controllers when the resources are reconciled successfully?

Reply:
Added success events and logs for each controller.